Python/tier2_redundancy_eliminator_bytecodes.c

```diff
@@ -189,6 +177,58 @@ dummy_func(void) {
         }
     }
 
+    op(_TO_BOOL, (value -- res)) {
+        (void)value;
+        OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+    }
+
+    op(_TO_BOOL_BOOL, (value -- value)) {
+        if (sym_matches_type(value, &PyBool_Type)) {
+            REPLACE_OP(this_instr, _NOP, 0, 0);
+        }
+        else {
+            sym_set_type(value, &PyBool_Type);
+        }
+    }
+
+    op(_TO_BOOL_INT, (value -- res)) {
+        sym_set_type(value, &PyLong_Type);
+        if (is_const(value)) {
+            PyObject *load = _PyLong_IsZero((PyLongObject *)get_const(value))
+                             ? Py_False : Py_True;
+            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)load);
+            OUT_OF_SPACE_IF_NULL(res = sym_new_const(ctx, load));
+        }
+        else {
+            OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+        }
+    }
+
+    op(_TO_BOOL_LIST, (value -- res)) {
+        sym_set_type(value, &PyList_Type);
+        OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+    }
+
+    op(_TO_BOOL_NONE, (value -- res)) {
+        if (get_const(value) == Py_None) {
+            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)Py_None);
+        }
+        sym_set_type(value, &_PyNone_Type);
```

Comment:
I'd move this up to the top of the block (but after the check for "impossible" cases, if you decide to add that).

Reply:
Traditionally, the rest of the code sets types after the operation. The thinking is that the types are known after the operation executes.

Reply:
Oh, subtle. I thought that this was an assertion about the input, and it is, but it's not actually known until the op has executed successfully -- it deopts if the input isn't None. So you are right.